### PR TITLE
Feat/60 add use intersection observer hook

### PR DESCRIPTION
**Pull request by francoisdillinger**
# Pull Request Description

**Description**
A hook that observes changes in the intersection of a target element with viewport.

**Acceptance Criteria**
- [x] **Element Visibility:** Detects when the target element enters or exits the viewport.
- [x] **Threshold/Margin Configuration:** Allows configuration thresholds/rootMargin.
- [x] **Animate Once:** Add the option to only animate the observed element only once on scroll.
- [x] **Cleanup:** Properly disconnects the observer when the component is unmounted.

**Details:**
- The `useIntersectionObserver` hook detects when a target element enters or exits the viewport.
- It allows configuration of thresholds and root margins for intersection detection.
- Includes an option to animate the observed element only once on scroll.
- Properly disconnects the observer when the component is unmounted.
- Updated the `index.ts` file to export the `useIntersectionObserver` hook.
- Updated the `docs/README` with a description of the new hook.

**Demo:**

Here is a gif demonstrating the hook in action:

Result when passed the following options: ```const options = {observerOptions:{threshold: 1, rootMargin: "0px"}};```
one

Result when passed the following options: ``` const options = {animateOnce: true, observerOptions:{threshold: 1, rootMargin: "0px"}};```
two

Result when passed the following options: ```const options = {}; ```
three

Result when passed the following options: ``` const options = {animateOnce: false, observerOptions:{threshold: 0, rootMargin: "200px"}};```
four

### Code Changes

**Hook Implementation:**
```javascript
import { useEffect, useState, useRef } from "react";

type IntersectionObserverOptions = {
  threshold?: number;
  rootMargin?: string;
};

type UseIntersectionObserverProps = {
  animateOnce?: boolean;
  options?: IntersectionObserverOptions;
};

type IntersectionObserverResult = {
  observeRef: React.MutableRefObject<Element | null>;
  isVisible: boolean;
}

export const useIntersectionObserver = ({
  animateOnce = false,
  options = {}
}: UseIntersectionObserverProps = {}): IntersectionObserverResult => {
  const observeRef = useRef<Element | null>(null);
  const [isVisible, setIsVisible] = useState(false);

  useEffect(() => {
    const currentRef = observeRef.current;
    if (!currentRef) return;

    const observer = new IntersectionObserver((entries) => {
      const entry = entries[0];
      if (!entry) return;

      setIsVisible(entry.isIntersecting);

      if (entry.isIntersecting && animateOnce) {
        observer.disconnect();
      }
    }, options);

    observer.observe(currentRef);

    return () => {
      observer.disconnect();
    };
  }, []);

  return { observeRef, isVisible };
};
```

**Index File Update:**
```typescript
export { useIntersectionObserver } from './hooks/useIntersectionObserver';
// ...other exports
```

**Documentation Update:**
Added a description of the `useIntersectionObserver` hook to the `docs/README` file.
```
- [`useIntersectionObserver()`](https://reacthaiku.dev/docs/hooks/useIntersectionObserver) - Provides a way to detect when an element enters or exits the viewport, with options for configuring intersection thresholds, margins, and one-time animation triggers.

### CODE REVIEW

File: lib/hooks/useIntersectionObserver.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  isVisible: boolean;
+}
+
+export const useIntersectionObserver = ({
+  animateOnce = false,
+  options = {}
+}: UseIntersectionObserverProps = {}): IntersectionObserverResult => {
+  const observeRef = useRef<Element | null>(null);
+  const [isVisible, setIsVisible] = useState(false);
+
```
Comment (DavidHDev) [inline lib/hooks/useIntersectionObserver.ts:24]: We should probably add a check for browser compatibility here, to return early if the IntersectionObserver API is not supported:

```typescript
    if (typeof IntersectionObserver === "undefined") {
      console.warn("IntersectionObserver is not supported in this browser.");
      return;
    }
```

File: lib/hooks/useIntersectionObserver.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        observer.disconnect();
+      }
+    }, options);
+
+    observer.observe(currentRef);
+
+    return () => {
+      observer.disconnect();
+    };
+  }, []);
```
Comment (DavidHDev) [inline lib/hooks/useIntersectionObserver.ts:45]: This empty dependency array can lead to issues if options or animateOnce change, as the effect will not re-run to adjust the observer settings, which you might want to dynamically change based on different requirements of the component it's being used in, so I think you can add these to the dependency array:

```
[observeRef, animateOnce, options]
```

File: lib/hooks/useIntersectionObserver.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import { useEffect, useState, useRef } from "react";
+
+type IntersectionObserverOptions = {
+  threshold?: number | number[];
```
Comment (francoisdillinger) [inline lib/hooks/useIntersectionObserver.ts:4]: Added `number[]` option for `threshold?` as I overlooked that in the documentation.

File: lib/hooks/useIntersectionObserver.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  animateOnce = false,
+  options = {}
+}: UseIntersectionObserverProps = {}): IntersectionObserverResult => {
+  const observeRef = useRef<Element | null>(null);
+  const [isVisible, setIsVisible] = useState(false);
+
+  if (typeof IntersectionObserver === "undefined") {
+    console.warn("IntersectionObserver is not supported in this browser.");
+    return { observeRef, isVisible: true };
+  }
```
Comment (francoisdillinger) [inline lib/hooks/useIntersectionObserver.ts:28]: Updated the dependency array

File: lib/hooks/useIntersectionObserver.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      }
+    }, options);
+
+    observer.observe(currentRef);
+
+    return () => {
+      observer.disconnect();
+    };
+  }, [observeRef, animateOnce, options]);
+
```
Comment (francoisdillinger) [inline lib/hooks/useIntersectionObserver.ts:51]: Updated the dependency array.

**Review by DavidHDev — APPROVED** [review]

**Comment by DavidHDev** [comment]
Looks good, thanks!

**Pull request merged**